core/reflection.jvm/src/kotlin/reflect/jvm/internal/KCallableImpl.kt

```diff
@@ -112,65 +112,87 @@ internal abstract class KCallableImpl<out R> : KCallable<R>, KTypeParameterOwner
         return if (isAnnotationConstructor) callAnnotationConstructor(args) else callDefaultMethod(args, null)
     }
 
+    private val _absentArguments = ReflectProperties.lazySoft {
+        val parameterSize = parameters.size + (if (isSuspend) 1 else 0)
+        val maskSize = (parameters.size + Integer.SIZE - 1) / Integer.SIZE
+
+        // arguments with masks and DefaultConstructorMarker or MethodHandle
+        // +1 is argument for DefaultConstructorMarker or MethodHandle
+        val arguments = arrayOfNulls<Any?>(parameterSize + maskSize + 1)
+
+        // set absent values
+        parameters.forEach { parameter ->
+            if (parameter.isOptional && !parameter.type.isInlineClassType) {
+                // For inline class types, the javaType refers to the underlying type of the inline class,
+                // but we have to pass null in order to mark the argument as absent for InlineClassAwareCaller.
+                arguments[parameter.index] = defaultPrimitiveValue(parameter.type.javaType)
+            } else if (parameter.isVararg) {
+                arguments[parameter.index] = defaultEmptyArray(parameter.type)
+            }
+        }
+
+        for (i in 0 until maskSize) {
+            arguments[parameterSize + i] = 0
+        }
+
+        arguments
+    }
+
+    private fun getAbsentArguments(): Array<Any?> = _absentArguments().clone()
+
     // See ArgumentGenerator#generate
     internal fun callDefaultMethod(args: Map<KParameter, Any?>, continuationArgument: Continuation<*>?): R {
         val parameters = parameters
-        val arguments = ArrayList<Any?>(parameters.size)
-        var mask = 0
-        val masks = ArrayList<Int>(1)
-        var index = 0
-        var anyOptional = false
+        val parameterSize = parameters.size + (if (isSuspend) 1 else 0)
```

Comment:
The following optimization to no-argument functions was found to be nearly twice as fast as 638d89c.
Does this optimization look like it should be incorporated?

```
Benchmark                          Mode  Cnt         Score         Error  Units
Measurement.zero                  thrpt    4  22299885.316 ±  295995.533  ops/s
Measurement.fiveWithDefault       thrpt    4   1927106.356 ±  225032.224  ops/s
Measurement.fiveWithoutDefault    thrpt    4   2460623.502 ±   96649.714  ops/s
Measurement.oneWithDefault        thrpt    4   5384826.666 ±  340254.100  ops/s
Measurement.oneWithoutDefault     thrpt    4   5654229.984 ± 5914104.417  ops/s
Measurement.twentyWithDefault     thrpt    4    510341.805 ±    2940.113  ops/s
Measurement.twentyWithoutDefault  thrpt    4    739231.653 ±   21298.002  ops/s
```

```suggestion
        val parameterSize = parameters.size + (if (isSuspend) 1 else 0)

        // Optimization for general no-argument functions.        
        if (parameters.isEmpty()) {
            @Suppress("UNCHECKED_CAST")
            return reflectionCall {
                caller.call(if (isSuspend) arrayOf(continuationArgument) else emptyArray()) as R
            }
        } else if (parameters.size == 1) {
            parameters.first().takeIf { it.kind != KParameter.Kind.VALUE }?.let { parameter ->
                @Suppress("UNCHECKED_CAST")
                return reflectionCall {
                    caller.call(if (isSuspend) arrayOf(args[parameter], continuationArgument) else arrayOf(args[parameter])) as R
                }
            }
        }
```

Reply:
For the 0-parameters case, it looks good! But for the 1 parameter, I'm not really sure, the code is getting quite complicated already. Besides, `Measurement.oneWithoutDefault` regressed in your results after this change, if I understand them correctly? I propose to keep the 0-parameter case only.

Reply:
> Measurement.oneWithoutDefault regressed in your results after this change, if I understand them correctly?

I ran the benchmark again with almost the same content, but there did not seem to be any regression in performance.
Since the benchmark was run on a local machine, there must have been something going on in the background during this measurement.

**1e1cd3a**

```
Benchmark                          Mode  Cnt         Score         Error  Units
Measurement.fiveWithDefault       thrpt    4   1713974.649 ±  807978.785  ops/s
Measurement.fiveWithoutDefault    thrpt    4   2402911.282 ±  116905.232  ops/s
Measurement.oneWithDefault        thrpt    4   5018352.177 ± 3050333.309  ops/s
Measurement.oneWithoutDefault     thrpt    4   6338081.662 ±  384633.083  ops/s
Measurement.twentyWithDefault     thrpt    4    506022.842 ±   28381.148  ops/s
Measurement.twentyWithoutDefault  thrpt    4    721987.591 ±   78852.517  ops/s
Measurement.zero                  thrpt    4  12727105.089 ±  188029.864  ops/s
```

**after-zero-arg-opt2**

```
Benchmark                          Mode  Cnt         Score         Error  Units
Measurement.fiveWithDefault       thrpt    4   1880153.629 ±  232291.891  ops/s
Measurement.fiveWithoutDefault    thrpt    4   2388820.000 ±   97785.718  ops/s
Measurement.oneWithDefault        thrpt    4   5276216.078 ±  327025.056  ops/s
Measurement.oneWithoutDefault     thrpt    4   6251850.821 ±  497046.067  ops/s
Measurement.twentyWithDefault     thrpt    4    516927.278 ±   13448.841  ops/s
Measurement.twentyWithoutDefault  thrpt    4    728619.156 ±   36904.744  ops/s
Measurement.zero                  thrpt    4  21774799.243 ± 1534253.147  ops/s
```

---

I agree that it is complicated.
I have pushed for a fix.
[1cf81dc](https://github.com/JetBrains/kotlin/pull/4840/commits/1cf81dc53638938a15f3cbe905562693ba2958d7)

I did not measure benchmark results with the one-parameter branch removed, but at least there is no reason for the score to drop since the target of the measurement is a top-level function.

Reply:
As a side note, the reason I proposed the original form is that `getter` is a typical example of a function that takes only instances as arguments.